### ISSUE TITLE
For click it's not clear how long to wait for a navigation to happen

### ISSUE THREAD

**Issue by whimboo** [w3c/webdriver#1046]
The current spec says the following for `click`:

> 9. Perform implementation-defined steps to allow any navigations triggered by the click to start.
> 10. Try to wait for navigation to complete. 

We can perform those implementation-defined steps but it's not clear how long it will take until the navigation starts. It can be from some milliseconds up to more then 100ms. It all depends on the CPU load, performance of the machine, and type of build of the browser (debug builds are generally slower). 

Also not each `click` triggers a navigation, which would mean that we loose a given number of milliseconds each time a `click` is not causing a navigation.

How can the spec cover that, or don't we want to do that? 

For Marionette I implemented that we are waiting for `beforeunload` events which are send first. So a hard-coded 200ms delay works just fine in all cases. But even then a navigation request can be prevented by a custom listener registered for `beforeunload`, or something else delays it.

Given that navigation is optional for `click` how can we define the duration the client has to wait? Also to not affect those `click` commands that much, which don't cause a navigation.

**Comment by shs96c** [comment]
TL;DR: it's implementation-specific.

This was discussed by @jgraham, @AutomatedTester, @andreastt, and myself when we met at Mozilla on Wednesday.

The implementation-specific steps might be as simple as putting a "tracing" event into the internal message queue after the actions required by the command and waiting until that is processed. This is the way the old IEDriver used to work. It is assumed that once those events have all been processed, some implementation-specific work will be required to determine whether or not a page load is expected. However the need for navigation is determined, it will need to be implemented differently per browser.

As an example, the XPI-based FirefoxDriver would compare the current and target URLs in certain circumstances, and would wait for navigation to start by registering an `nsIProgressListener`.

How long should we wait until "navigation" starts? That depends on how the implementation-specific steps are implemented. 

We know that the mechanisms used are inherently racy, which is why usage of webdriver normally comes with encouragement to use explicit waits, and why we added implicit waits too.
